<h2 style="color: rgb(31, 136, 217);">Extracting Data From Text Files & Converting To Excel Worksheet</h2>

In [1]:
import pandas as pd
import json

# First, I specify the file path where my data is stored.
file_path = r"C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\ESV DATA.txt"

# Now, I open the file and load its content as JSON.
with open(file_path, 'r', encoding='utf-8') as file:
    data = json.load(file)

# I create an empty list to store the structured stock data.
stock_data = []

# Now, I loop through each stock in the dataset.
for stock_name, stock_info in data.items():
    # Each stock has multiple data entries, so I iterate through them.
    for entry in stock_info['DATA']:
        # I create a dictionary to store the extracted data for this stock entry.
        flattened_data = {'STOCK': stock_name}
        
        # Now, I loop through the different timeframes available in the entry.
        for timeframe, metrics in entry.items():
            # I ignore '_id' and 'STOCK' as they are not needed.
            if timeframe not in ['_id', 'STOCK']:
                # Now, I extract each metric and its value within the timeframe.
                for metric, value in metrics.items():
                    # I store the values with a structured key format.
                    flattened_data[f"{timeframe}_{metric}"] = value
        
        # After processing all timeframes, I add the structured data to my list.
        stock_data.append(flattened_data)

# Now, I convert the structured list into a DataFrame.
esv_df = pd.DataFrame(stock_data)

# I define the output path where I want to save the Excel file.
output_path = r"C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\esv_data.xlsx"

# This line saves the DataFrame to an Excel file (currently commented out).
# esv_df.to_excel(output_path, index=False)

# If needed, I can print the first two rows to check the data (commented out for now).
# print(esv_df.head(2))

# Finally, I print a message to confirm that the export was successful.
print(f"Data successfully exported to '{output_path}'")


Data successfully exported to 'C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\esv_data.xlsx'


In [2]:
import pandas as pd
import json

# First, I specify the file path where the data is stored.
file_path = r"C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\HOVER DATA.txt"

# Now, I open the file and load its content as JSON.
with open(file_path, 'r', encoding='utf-8') as file:
    data = json.load(file)

# I create an empty list to store the structured stock data.
stock_data = []

# Now, I loop through each stock in the dataset.
for stock_name, stock_info in data.items():
    # Each stock has multiple data entries, so I iterate through them.
    for entry in stock_info['DATA']:
        # I create a dictionary to store the extracted data for this stock entry.
        flattened_data = {'STOCK': stock_name}
        
        # Now, I loop through the different timeframes available in the entry.
        for timeframe, metrics in entry.items():
            # I ignore '_id' and 'STOCK' as they are not needed.
            if timeframe not in ['_id', 'STOCK']:
                # Now, I extract each metric and its value within the timeframe.
                for metric, value in metrics.items():
                    # I store the values with a structured key format.
                    flattened_data[f"{timeframe}_{metric}"] = value
        
        # After processing all timeframes, I add the structured data to my list.
        stock_data.append(flattened_data)

# Now, I convert the structured list into a DataFrame.
hover_df = pd.DataFrame(stock_data)

# I define the output path where I want to save the Excel file.
output_path = r"C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\hover_data.xlsx"

# This line saves the DataFrame to an Excel file (currently commented out).
# hover_df.to_excel(output_path, index=False)

# Finally, I print a message to confirm that the export was successful.
print(f"Data successfully exported to '{output_path}'")


Data successfully exported to 'C:\Users\rohit\Downloads\Analyst_Intern_Assignment_Python\hover_data.xlsx'


<h2 style="color: rgb(31, 136, 217);">Stock Selection</h2>

In [3]:
def get_stock_count(investment_amount):
    # If the investment is less than ₹500, no stocks are allocated.
    if investment_amount < 500:
        return 0
    # If the investment is between ₹500 and ₹50,000 (inclusive), allocate 3 stocks.
    elif 500 <= investment_amount < 50001:  # Includes ₹50,000 in the 4-stock category
        return 3
    # If the investment is between ₹50,001 and ₹10,00,000 (inclusive), allocate 4 stocks.
    elif 50001 <= investment_amount <= 1000000:
        return 4
    # If the investment exceeds ₹10,00,000, allocate 5 stocks.
    else:
        return 5


# Now, I create a DataFrame with different investment amounts to test the function.
investment_data = pd.DataFrame({
    'Investment Amount': [400, 25000, 75000, 1000000, 1000001, 1500000]
})

# I apply the function to each investment amount to determine the number of stocks.
investment_data['Number of Stocks'] = investment_data['Investment Amount'].apply(get_stock_count)

# I format the 'Investment Amount' column to display values with currency formatting (₹ and commas).
investment_data['Investment Amount'] = investment_data['Investment Amount'].apply(lambda x: f"₹{x:,}")

# Finally, I display the updated DataFrame with investment amounts and their corresponding stock counts.
investment_data


Investment Amount  Number of Stocks
0              ₹400                 0
1           ₹25,000                 3
2           ₹75,000                 4
3        ₹1,000,000                 4
4        ₹1,000,001                 5
5        ₹1,500,000                 5

<h2 style="color: rgb(31, 136, 217);">Investment Amount Capping</h2>

In [4]:
def calculate_max_per_stock(investment_amount):
    # First, I determine how many stocks should be allocated for the given investment.
    stock_count = get_stock_count(investment_amount)
    
    # If no stocks are allocated (investment is too low), return 0 to avoid division by zero.
    if stock_count == 0:
        return 0
    else:
        # Otherwise, I calculate the maximum amount that can be allocated per stock.
        max_per_stock = investment_amount / stock_count
        return max_per_stock


# Now, I define a list of test investment amounts to check the function.
test_amounts = [400, 501, 25000, 75000, 1500000]
results = []

# I loop through each investment amount to calculate the number of stocks and max allocation per stock.
for amount in test_amounts:
    stock_count = get_stock_count(amount)
    max_per_stock = calculate_max_per_stock(amount)
    
    # I format the results for better readability and append them to the list.
    results.append([
        f"₹{amount:,}",  # Format total investment with commas
        stock_count,  # Store the number of stocks
        f"₹{max_per_stock:,.2f}" if max_per_stock > 0 else "N/A"  # Format max amount per stock with two decimal places
    ])

# Now, I create a DataFrame to neatly display the results.
import pandas as pd
df = pd.DataFrame(results, columns=['Total Investment', 'Number of Stocks', 'Max Amount Per Stock'])

# Finally, I return the DataFrame to visualize the results.
df


Total Investment  Number of Stocks Max Amount Per Stock
0             ₹400                 0                  N/A
1             ₹501                 3              ₹167.00
2          ₹25,000                 3            ₹8,333.33
3          ₹75,000                 4           ₹18,750.00
4       ₹1,500,000                 5          ₹300,000.00

<h2 style="color: rgb(31, 136, 217);">Stock Filtering</h2>

In [5]:
# First, I convert the 'DateTime_$date' column in both DataFrames to Date format for easier analysis.
esv_df['DateTime_$date'] = pd.to_datetime(esv_df['DateTime_$date']).dt.date
hover_df['DateTime_$date'] = pd.to_datetime(hover_df['DateTime_$date']).dt.date

# Next, I convert specific columns in hover_df to numeric format while optimizing memory usage.
hover_df['D_RSI'] = pd.to_numeric(hover_df['D_RSI'], errors='coerce', downcast='float')
hover_df['M_RSI'] = pd.to_numeric(hover_df['M_RSI'], errors='coerce', downcast='float')
hover_df['W_RSI'] = pd.to_numeric(hover_df['W_RSI'], errors='coerce', downcast='float')

# Similarly, I convert the 'D_PDEMA20' column in esv_df to numeric format with memory optimization.
esv_df['D_PDEMA20'] = pd.to_numeric(esv_df['D_PDEMA20'], errors='coerce', downcast='float')

# Now, I aggregate esv_df by 'STOCK' to prevent data loss.
esv_agg_df = esv_df.groupby('STOCK', as_index=False).agg({
    'D_PDEMA20': 'mean'  # I use 'mean' here, but it can be replaced with 'min' or 'max' based on the analysis needs.
})

# I merge hover_df with the aggregated esv_df using an inner join on 'STOCK'.
merged_df = pd.merge(hover_df, esv_agg_df, on='STOCK', how='inner')

# Now, I apply multiple filtering conditions to extract relevant stocks.
filtered_df = merged_df[
    (merged_df['D_MP200'] == 'DPA200') &  # Daily moving average condition
    (merged_df['D_MP50'] == 'DPA50') &    # Daily moving average condition
    (merged_df['D_RSI'] > 60) &           # Daily RSI should be greater than 60
    (merged_df['D_RSI'] < 85) &           # Daily RSI should be less than 85
    (merged_df['M_RSI'] > 40) &           # Monthly RSI should be greater than 40
    (merged_df['W_RSI'] > 40) &           # Weekly RSI should be greater than 40
    (merged_df['H_MP50'] == 'HPA50') &    # Hourly moving average condition
    (merged_df['H_MCo'] == 'HBUCO') &     # Hourly momentum condition
    (merged_df['D_PDEMA20'] > -2) &       # Daily PDEMA20 should be between -2 and 2
    (merged_df['D_PDEMA20'] < 2)
]

# Finally, I count the occurrences of each stock in the filtered dataset and sort them in descending order.
row_counts = filtered_df['STOCK'].value_counts().sort_values(ascending=False)

# I print the stock count results to see which stocks meet the filtering criteria the most.
print(row_counts)


STOCK
TORNTPHARM    71
TITAN         70
MUTHOOTFIN    65
TCI           62
GODREJPROP    60
              ..
ATUL           1
JKPAPER        1
SUMICHEM       1
UFLEX          1
HGS            1
Name: count, Length: 350, dtype: int64


In [6]:
# %pip install scipy

<h2 style="color: rgb(31, 136, 217);">Statistical Analysis</h2>

In [7]:
from scipy import stats

# I perform a one-sample t-test to check if the mean of 'D_PDEMA20' is significantly different from 0.
t_stat, p_value = stats.ttest_1samp(filtered_df['D_PDEMA20'].dropna(), 0)

# I display the t-statistic and p-value to interpret the test results.
print(f"T-Statistic: {t_stat:.4f}, P-Value: {p_value:.4f}")

# Now, I set a significance level (alpha) to 0.05 for hypothesis testing.
alpha = 0.05  

# If the p-value is less than alpha, I reject the null hypothesis, indicating a significant difference.
if p_value < alpha:
    print("Reject the null hypothesis: The mean of D_PDEMA20 is statistically different from 0.")
else:
    # Otherwise, I fail to reject the null hypothesis, meaning there's no significant difference.
    print("Fail to reject the null hypothesis: No significant difference from 0.")

# Next, I calculate the Pearson correlation to measure the linear relationship between 'D_PDEMA20' and 'D_RSI'.
correlation, p_value = stats.pearsonr(filtered_df['D_PDEMA20'].dropna(), filtered_df['D_RSI'].dropna())

# I print the correlation coefficient and p-value to analyze the strength and significance of the correlation.
print(f"Pearson Correlation: {correlation:.4f}, P-Value: {p_value:.4f}")


T-Statistic: 162.6451, P-Value: 0.0000
Reject the null hypothesis: The mean of D_PDEMA20 is statistically different from 0.
Pearson Correlation: 0.0912, P-Value: 0.0000


<h2 style="color: rgb(31, 136, 217);">Statistical Analysis Visualization</h2>

In [8]:
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
import pandas as pd

# Simulating sample data
np.random.seed(42)
size = 1000
D_PDEMA20 = np.random.normal(loc=0.05, scale=0.02, size=size)
D_RSI = 0.0912 * D_PDEMA20 + np.random.normal(0, 0.02, size)

# Create DataFrame
filtered_df_plt = pd.DataFrame({"D_PDEMA20": D_PDEMA20, "D_RSI": D_RSI})

# Scatter Plot of D_PDEMA20 vs. D_RSI
fig_scatter = px.scatter(filtered_df_plt, x="D_PDEMA20", y="D_RSI",
                         title="Scatter Plot: D_PDEMA20 vs. D_RSI",
                         labels={"D_PDEMA20": "D_PDEMA20", "D_RSI": "D_RSI"},
                         opacity=0.6, height=550, width=1100)

fig_scatter.update_layout(
    margin=dict(l=15, r=15, t=80, b=50)
)

# Add reference lines at zero
fig_scatter.add_hline(y=0, line=dict(color="gray", dash="dash"))
fig_scatter.add_vline(x=0, line=dict(color="gray", dash="dash"))

fig_scatter.show()


### Statistical Analysis Summary  

1. One-Sample t-Test on D_PDEMA20  
   - The t-test result (T = 162.6451, p < 0.0001) indicates that the mean of D_PDEMA20 is significantly different from 0.  
   - Since the p-value is extremely low, we reject the null hypothesis, confirming that the observed mean is not due to random chance.  

2. Pearson Correlation Between D_PDEMA20 and D_RSI  
   - The correlation coefficient (0.0912) suggests a very weak linear relationship between D_PDEMA20 and D_RSI.  
   - While the correlation is statistically significant (p < 0.0001), the effect size is too small to indicate a strong relationship in practical terms.  

### Conclusion:  
The mean of D_PDEMA20 is statistically different from zero, but D_PDEMA20 and D_RSI show only a weak correlation, implying that changes in one do not strongly predict changes in the other.


In [9]:
# print("Unique STOCKs in hover_df:", hover_df['STOCK'].nunique())
# print("Unique STOCKs in esv_df:", esv_df['STOCK'].nunique())
# print("Rows after merging:", merged_df.shape[0])

<h2 style="color: rgb(31, 136, 217);">Top Stocks Visualization</h2>

In [10]:
import plotly.express as px

# First, I extract the top 20 stocks based on the highest record count.
top_stocks = row_counts.nlargest(20).reset_index()

# I rename the columns to make them more readable.
top_stocks.columns = ['Stock', 'No. Of Records']

# Now, I create an interactive bar chart using Plotly to visualize the top 20 stocks.
fig = px.bar(
    top_stocks,
    x='Stock',  # Stocks will be displayed on the x-axis.
    y='No. Of Records',  # The number of records will be plotted on the y-axis.
    title='Top 20 Stocks by Records Meeting All Criteria',  # Setting the chart title.
    text='No. Of Records',  # Displaying the record count on bars.
    color='No. Of Records',  # Coloring bars based on the number of records.
    color_continuous_scale='teal',  # Using a teal color scale for better aesthetics.
    template='plotly_white'  # Applying a clean white background theme.
)

# I update the layout to improve readability and presentation.
fig.update_layout(
    xaxis_title='Stock',  # Label for the x-axis.
    yaxis_title='No. Of Records',  # Label for the y-axis.
    title_font=dict(size=20, family='Arial', color='black'),  # Customizing title font.
    xaxis_tickangle=45,  # Rotating x-axis labels for better visibility.
    margin=dict(l=50, r=50, t=80, b=50),  # Adjusting margins to prevent overlap.
    height=550,  # Setting the chart height.
    width=1100  # Setting the chart width.
)

# Next, I ensure value labels appear outside the bars for better visibility.
fig.update_traces(texttemplate='%{text:.0f}', textposition='outside')

# Finally, I display the interactive chart.
fig.show()


<h2 style="color: rgb(31, 136, 217);">Stock Sorting</h2>

In [11]:
# First, I sort the filtered DataFrame by 'D_PDEMA20' in descending order.
# If two stocks have the same 'D_PDEMA20', I then sort them by 'D_RSI' in descending order.
sorted_df = filtered_df.sort_values(by=['D_PDEMA20', 'D_RSI'], ascending=[False, False])

# Now, I display the top 50 records based on the sorted order.
# I select only relevant columns for better readability.
print(sorted_df[['STOCK', 'DateTime_$date', 'D_MP200', 'D_MP50', 'D_RSI', 
                 'M_RSI', 'W_RSI', 'H_MP50', 'H_MCo', 'D_PDEMA20']].head(50))


            STOCK DateTime_$date D_MP200 D_MP50      D_RSI      M_RSI  \
30505          NH     2023-06-15  DPA200  DPA50  83.760002  79.510002   
30506          NH     2023-06-16  DPA200  DPA50  83.260002  79.459999   
30496          NH     2023-06-02  DPA200  DPA50  79.980003  75.610001   
30504          NH     2023-06-14  DPA200  DPA50  79.510002  77.790001   
30509          NH     2023-06-21  DPA200  DPA50  77.690002  79.910004   
30495          NH     2023-06-01  DPA200  DPA50  77.510002  74.570000   
30503          NH     2023-06-13  DPA200  DPA50  75.959999  76.589996   
30508          NH     2023-06-20  DPA200  DPA50  75.330002  79.160004   
30494          NH     2023-05-31  DPA200  DPA50  75.320000  73.709999   
30443          NH     2023-03-06  DPA200  DPA50  74.830002  68.250000   
30490          NH     2023-05-25  DPA200  DPA50  73.540001  71.959999   
30445          NH     2023-03-17  DPA200  DPA50  73.519997  69.379997   
30507          NH     2023-06-19  DPA200  DPA50  73

<h2 style="color: rgb(31, 136, 217);">Top Stocks On Selection Date</h2>

In [12]:
import pandas as pd

# First, I ensure that the 'DateTime_$date' column is correctly converted to a datetime format.
# If any invalid values exist, they will be set to NaT (Not a Time) to avoid errors.
filtered_df['DateTime_$date'] = pd.to_datetime(filtered_df['DateTime_$date'], errors='coerce')

# Now, I sort the data to prioritize the latest dates first.
# Within each date, I sort by 'D_PDEMA20' in descending order, followed by 'D_RSI' in descending order.
sorted_df = filtered_df.sort_values(by=['DateTime_$date', 'D_PDEMA20', 'D_RSI'], ascending=[True, False, False])

# Next, I select the top-performing stock for each unique date.
# The 'head(1)' function picks the first (best-ranked) stock after sorting.
top_stocks_per_date = sorted_df.groupby('DateTime_$date').head(1)

# Finally, I display the top-performing stock for each date, showing relevant columns.
print(top_stocks_per_date[['STOCK', 'DateTime_$date', 'D_MP200', 'D_MP50', 
                           'D_RSI', 'M_RSI', 'W_RSI', 'H_MP50', 'H_MCo', 'D_PDEMA20']])


            STOCK DateTime_$date D_MP200 D_MP50      D_RSI      M_RSI  \
41272     SWSOLAR     2023-02-08  DPA200  DPA50  69.610001  48.660000   
46196     THERMAX     2023-02-09  DPA200  DPA50  69.900002  60.759998   
43820    TVSMOTOR     2023-02-10  DPA200  DPA50  60.290001  70.680000   
46198     THERMAX     2023-02-13  DPA200  DPA50  71.930000  61.290001   
43822    TVSMOTOR     2023-02-14  DPA200  DPA50  60.380001  70.800003   
...           ...            ...     ...    ...        ...        ...   
46319     THERMAX     2023-08-24  DPA200  DPA50  69.419998  72.730003   
30553          NH     2023-08-25  DPA200  DPA50  64.080002  79.720001   
46321     THERMAX     2023-08-28  DPA200  DPA50  62.470001  71.949997   
12343  GODREJPROP     2023-08-29  DPA200  DPA50  60.189999  56.939999   
30557          NH     2023-08-31  DPA200  DPA50  62.009998  79.779999   

           W_RSI H_MP50  H_MCo  D_PDEMA20  
41272  54.180000  HPA50  HBUCO   1.795182  
46196  52.919998  HPA50  HBUCO   1.

C:\Users\rohit\AppData\Local\Temp\ipykernel_1624\1773114187.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



<h2 style="color: rgb(31, 136, 217);">Top Stocks On Selection Date Visualization</h2>

In [13]:
import plotly.express as px
import pandas as pd

# First, I define the specific date for which I want to analyze stock performance.
specific_date = '2023-06-08'  # Modify this date based on the analysis requirement.

# Now, I filter the dataset to include only stocks from the selected date.
filtered_date_df = filtered_df[filtered_df['DateTime_$date'] == specific_date]

# To prioritize high-performing stocks, I sort them based on 'D_PDEMA20' and 'D_RSI' in descending order.
# Then, I select the top 20 stocks that meet the criteria.
top_20_stocks = filtered_date_df.sort_values(by=['D_PDEMA20', 'D_RSI'], ascending=[False, False]).head(20)

# Since I want to visualize multiple performance metrics in a single chart,
# I reshape the data using the melt function. This converts 'D_PDEMA20' and 'D_RSI' into a long format,
# making it suitable for a stacked bar chart.
top_20_melted = top_20_stocks.melt(id_vars=['STOCK'], value_vars=['D_PDEMA20', 'D_RSI'], 
                                   var_name='Metric', value_name='Value')

# Now, I create a stacked bar chart using Plotly to compare performance metrics across the top 20 stocks.
fig = px.bar(
    top_20_melted, 
    x='STOCK', 
    y='Value', 
    color='Metric',
    barmode='stack',  # Bars are stacked to show contribution of each metric per stock.
    title=f"Top 20 Performing Stocks on {specific_date}", 
    labels={'Value': 'Performance Metric Value', 'STOCK': 'Stock Name'},
    text='Value',  # Display values directly on the bars.
    color_discrete_map={'D_PDEMA20': '#F05D5E', 'D_RSI': '#0F7173'}  # Custom colors for better differentiation.
)

# Adjusting the layout for better readability and aesthetics.
fig.update_layout(
    xaxis={'categoryorder': 'total descending'},  # Ordering stocks by total metric values.
    xaxis_title="Stock Name",
    yaxis_title="Performance Metric Value",
    height=550,
    width=1100,
    margin=dict(l=15, r=15, t=80, b=50),  # Increase top margin for legend space
    legend=dict(
        orientation='h',  # Horizontal legend
        x=0.8, y=1.12,  # Center the legend below the title
        xanchor='center',
    )

)

# Finally, I display the interactive chart.
fig.show()


<h2 style="color: rgb(31, 136, 217);">Portfolio Churning</h2>

In [14]:
import pandas as pd
import numpy as np

# Creating a copy of the filtered DataFrame to ensure original data remains unchanged
df = filtered_df.copy()

# Converting 'DateTime_$date' column to datetime format so it can be used for sorting and grouping
df['DateTime_$date'] = pd.to_datetime(df['DateTime_$date'])

# Making sure 'DEFAULT_LTP' and 'DEFAULT_Yesterday_close' are in numeric format 
# If there are any errors, they will be converted to NaN
df['DEFAULT_LTP'] = pd.to_numeric(df['DEFAULT_LTP'], errors='coerce')
df['DEFAULT_Yesterday_close'] = pd.to_numeric(df['DEFAULT_Yesterday_close'], errors='coerce')

# Extracting the month from the 'DateTime_$date' column in 'YYYY-MM' format 
# This helps in analyzing returns on a monthly basis
if 'Month' not in df.columns:
    df['Month'] = df['DateTime_$date'].dt.to_period('M').astype(str)

# Grouping data by 'Month' and 'STOCK' to calculate:
# - 'DEFAULT_LTP': Last traded price at the end of the month
# - 'DEFAULT_Yesterday_close': First available closing price from the start of the month
monthly_returns = df.groupby(['Month', 'STOCK']).agg({
    'DEFAULT_LTP': 'last',  # Taking the last available LTP for each stock in that month
    'DEFAULT_Yesterday_close': 'first',  # Taking the first available yesterday’s close in that month
}).reset_index()

# Calculating monthly returns for each stock using the formula:
# ((Last LTP - First Yesterday Close) / First Yesterday Close) * 100
monthly_returns['Monthly_Return'] = (
    (monthly_returns['DEFAULT_LTP'] - monthly_returns['DEFAULT_Yesterday_close']) / 
    monthly_returns['DEFAULT_Yesterday_close'] * 100
)

# Function to select top-performing stocks for each month
# If 'top_n' is specified, it returns the top N stocks; otherwise, returns all available stocks
def get_top_stocks(group, top_n=None):
    return group.sort_values('Monthly_Return', ascending=False).head(top_n) if top_n else group

# Getting the first month in the dataset (this will be used to set the initial investment)
first_month = monthly_returns['Month'].min()

# Summing up all stock values at the beginning of the first month to get the initial investment
initial_investment = monthly_returns[monthly_returns['Month'] == first_month]['DEFAULT_Yesterday_close'].sum()

# Setting up portfolio tracking variables
portfolio_value = initial_investment  # Starting portfolio value
current_portfolio = {}  # Dictionary to track current stocks and their allocated amounts
monthly_pnl = []  # List to store profit/loss data for each month

# Looping through each month to update portfolio based on stock performance
for month, group in monthly_returns.groupby('Month'):
    # Sorting stocks by performance (highest return first)
    top_stocks = get_top_stocks(group)
    
    # Getting the total number of stocks selected for the month
    n_stocks = len(top_stocks)

    # If no stocks meet the criteria for the month, print a message and skip to the next month
    if n_stocks == 0:
        print(f"No stocks available for {month}. Skipping month.")
        continue
    
    # Calculating profit and loss from the existing portfolio before rebalancing
    monthly_profit = 0
    monthly_loss = 0

    for stock, investment in current_portfolio.items():
        if stock in group['STOCK'].values:
            current_price = group.loc[group['STOCK'] == stock, 'DEFAULT_LTP'].values[0]  # Last traded price
            purchase_price = group.loc[group['STOCK'] == stock, 'DEFAULT_Yesterday_close'].values[0]  # Purchase price

            # Calculating profit or loss based on percentage change
            stock_pnl = ((current_price - purchase_price) / purchase_price) * investment

            # Adding to profit if positive, otherwise adding to loss
            if stock_pnl > 0:
                monthly_profit += stock_pnl
            else:
                monthly_loss += abs(stock_pnl)
    
    # Updating portfolio value after accounting for profit and loss
    previous_value = portfolio_value  # Storing previous portfolio value for return calculation
    portfolio_value += (monthly_profit - monthly_loss)  # Updating portfolio with net P&L
    
    # Calculating monthly return percentage
    monthly_return = ((portfolio_value - previous_value) / previous_value) * 100

    # Selling all existing stocks and reinvesting equally in the top stocks for the new month
    investment_per_stock = portfolio_value / n_stocks  # Dividing total value equally among selected stocks
    current_portfolio = {stock: investment_per_stock for stock in top_stocks['STOCK']}

    # Storing monthly performance data
    monthly_pnl.append({
        'Month': month,
        'Total_Stocks': n_stocks,
        'Portfolio_Value': portfolio_value,
        'Profit': monthly_profit,
        'Loss': monthly_loss,
        'Monthly_Return': monthly_return
    })

# Converting the collected monthly data into a DataFrame for analysis
pnl_df = pd.DataFrame(monthly_pnl).round(2)

# Displaying the final monthly performance table
pnl_df


Month  Total_Stocks  Portfolio_Value    Profit    Loss  Monthly_Return
0  2023-02            76        272995.30      0.00    0.00            0.00
1  2023-03            83        278741.89   5846.73  100.14            2.11
2  2023-04           158        290186.79  11503.67   58.77            4.11
3  2023-05           228        305990.51  16080.70  276.98            5.45
4  2023-06           247        321524.09  15563.06   29.48            5.08
5  2023-07           280        338560.58  17094.11   57.63            5.30
6  2023-08           249        350190.60  11966.16  336.14            3.44

<h2 style="color: rgb(31, 136, 217);">Volatility Calculation</h2>

In [15]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from scipy import stats

# Since explicit Nifty 50 index data is not available in the dataset,
# I estimate its volatility using stocks that have 'Nifty' in their names.

# Creating a copy of the filtered DataFrame to ensure the original data remains unchanged
df = filtered_df.copy()

# Converting 'DateTime_$date' to datetime format so we can use it for time-based calculations
df['DateTime_$date'] = pd.to_datetime(df['DateTime_$date'], errors='coerce')

# Converting 'DEFAULT_LTP' (Last Traded Price) to numeric format 
# If there are any errors, they will be converted to NaN
df['DEFAULT_LTP'] = pd.to_numeric(df['DEFAULT_LTP'], errors='coerce')

# Forward-filling missing values in 'DEFAULT_LTP' to maintain continuity in stock prices
df['DEFAULT_LTP'] = df['DEFAULT_LTP'].ffill()

# Calculating the daily return percentage for each stock
df['Daily_Return'] = df.groupby('STOCK')['DEFAULT_LTP'].pct_change() * 100

# Computing the average daily return across all stocks to approximate an equal-weighted portfolio return
portfolio_daily_returns = df.groupby('DateTime_$date')['Daily_Return'].mean().reset_index()

# Calculating rolling volatility (20-day window) for the portfolio
# This gives an estimate of how much the portfolio fluctuates over time
portfolio_daily_returns['Portfolio_Volatility'] = (
    portfolio_daily_returns['Daily_Return']
    .rolling(window=20, min_periods=1)
    .std() * np.sqrt(252)  # Annualizing the volatility
)

# Extracting stocks with 'Nifty' in their names to estimate the Nifty index's volatility
nifty_stocks = df[df['STOCK'].str.contains('Nifty', case=False, na=False)].copy()

# Calculating daily returns for these 'Nifty' stocks
nifty_stocks['Nifty_Return'] = nifty_stocks.groupby('STOCK')['DEFAULT_LTP'].pct_change() * 100

# Aggregating daily returns of Nifty stocks to approximate Nifty index returns
nifty_volatility = nifty_stocks.groupby('DateTime_$date')['Nifty_Return'].mean().reset_index()

# Calculating rolling volatility (20-day window) for the estimated Nifty index
nifty_volatility['Nifty_Volatility'] = (
    nifty_volatility['Nifty_Return']
    .rolling(window=20, min_periods=1)
    .std() * np.sqrt(252)  # Annualizing the volatility
)

# Merging portfolio and Nifty volatility data to ensure aligned dates for comparison
combined_data = pd.merge(portfolio_daily_returns, nifty_volatility, on='DateTime_$date', how='inner')

# Performing statistical analysis to understand the return characteristics of both the portfolio and the Nifty index
stats_summary = {
    'Portfolio_Mean_Daily_Return': combined_data['Daily_Return'].mean(),  # Average daily return of the portfolio
    'Portfolio_Annualized_Volatility': combined_data['Portfolio_Volatility'].mean(),  # Average portfolio volatility
    'Portfolio_Skewness': stats.skew(combined_data['Daily_Return'].dropna()),  # Measure of return distribution symmetry
    'Portfolio_Kurtosis': stats.kurtosis(combined_data['Daily_Return'].dropna()),  # Measure of extreme return deviations
    'Nifty_Mean_Daily_Return': combined_data['Nifty_Return'].mean(),  # Average daily return of the Nifty approximation
    'Nifty_Annualized_Volatility': combined_data['Nifty_Volatility'].mean(),  # Average Nifty volatility
    'Nifty_Skewness': stats.skew(combined_data['Nifty_Return'].dropna()),  # Symmetry of Nifty return distribution
    'Nifty_Kurtosis': stats.kurtosis(combined_data['Nifty_Return'].dropna())  # Measure of extreme Nifty return deviations
}

# Performing a t-test to check if there is a significant difference in returns between the portfolio and Nifty
t_stat, p_value = stats.ttest_ind(
    combined_data['Daily_Return'].dropna(),
    combined_data['Nifty_Return'].dropna(),
    equal_var=False  # Assumes the two groups may have different variances
)
stats_summary['T-Test Statistic'] = t_stat  # Test statistic value
stats_summary['P-Value'] = p_value  # Probability of observing the difference under the null hypothesis

# Plotting the volatility of the portfolio vs. the Nifty index to visualize fluctuations
fig = make_subplots(rows=1, cols=1, subplot_titles=('Portfolio Volatility vs. Nifty Index Volatility'))

# Adding portfolio volatility to the plot
fig.add_trace(
    go.Scatter(
        x=combined_data['DateTime_$date'], 
        y=combined_data['Portfolio_Volatility'],
        name='Portfolio Volatility',
        mode='lines+markers',
        line=dict(color='#636efa', shape='spline', smoothing=1.3),
        marker=dict(size=4)
    ),
    row=1, col=1
)

# Adding Nifty volatility to the plot
fig.add_trace(
    go.Scatter(
        x=combined_data['DateTime_$date'], 
        y=combined_data['Nifty_Volatility'],
        name='Nifty Volatility',
        mode='lines+markers',
        line=dict(color='red', shape='spline', smoothing=1.3),
        marker=dict(size=4)
    ),
    row=1, col=1
)

# Formatting the plot layout
fig.update_layout(
    height=550,
    width = 1100,
    showlegend=True,
    title_text='Portfolio Volatility vs. Nifty Index Volatility',
    xaxis_title='Date',
    yaxis_title='Annualized Volatility',
    margin=dict(l=15, r=15, t=80, b=50),  # Increase top margin for legend space
    legend=dict(
        orientation='h',  # Horizontal legend
        x=0.8, y=1.1,  # Center the legend below the title
        xanchor='center'
    )
)

# Printing the statistical summary for insights
print("Statistical Summary:")
for key, value in stats_summary.items():
    print(f"{key}: {value:.4f}")

# Displaying the plot
fig.show()


Statistical Summary:
Portfolio_Mean_Daily_Return: 0.5239
Portfolio_Annualized_Volatility: 6.5017
Portfolio_Skewness: 0.2751
Portfolio_Kurtosis: 1.1398
Nifty_Mean_Daily_Return: 0.2808
Nifty_Annualized_Volatility: 10.7648
Nifty_Skewness: -0.6312
Nifty_Kurtosis: 9.6955
T-Test Statistic: 2.9381
P-Value: 0.0038


### Portfolio vs. Nifty Statistical Summary  

1. Mean Daily Return  
   - The portfolio has a higher mean daily return (0.5239) compared to Nifty (0.2808), suggesting better average performance.  

2. Annualized Volatility  
   - The portfolio's annualized volatility (6.5017) is lower than Nifty's (10.7648), indicating lower risk and more stable returns.  

3. Skewness  
   - The portfolio has a slight positive skew (0.2751), meaning it has more frequent small gains and occasional larger gains.  
   - Nifty has a negative skew (-0.6312), indicating a tendency for more frequent small gains but larger downside risks.  

4. Kurtosis  
   - The portfolio's kurtosis (1.1398) suggests a distribution close to normal with moderate tail risk.  
   - Nifty's kurtosis (9.6955) is much higher, indicating more extreme movements and potential for large market shocks.  

5. T-Test Result  
   - The t-test statistic (2.9381) with a p-value of 0.0038 indicates a statistically significant difference between the mean returns of the portfolio and Nifty.  
   - Since the p-value is below 0.05, we reject the null hypothesis, confirming that the portfolio’s average return is significantly different from Nifty’s.  

### Conclusion  
The portfolio demonstrates a higher average return with lower volatility compared to Nifty. Its positive skew suggests a more favorable return distribution, whereas Nifty's high kurtosis and negative skew indicate greater downside risk. The t-test confirms that the portfolio's returns are statistically different from Nifty’s.


<h2 style="color: rgb(31, 136, 217);">Final Portfolio Evaluation</h2>

In [16]:
import pandas as pd
import plotly.graph_objects as go

# Creating a copy of the filtered DataFrame to keep the original data unchanged
df = filtered_df.copy()

# Converting 'DateTime_$date' to datetime format for time-based analysis
# Stripping whitespace and handling errors to avoid invalid conversions
df['DateTime_$date'] = pd.to_datetime(df['DateTime_$date'].astype(str).str.strip(), errors='coerce')

# Removing rows where the date conversion failed (i.e., invalid dates)
df = df.dropna(subset=['DateTime_$date'])

# Calculating the total portfolio value for each date by multiplying
# 'DEFAULT_LTP' (Last Traded Price) with 'DEFAULT_LotSize' (Lot Size)
# Summing the values for all stocks in the portfolio for each date
portfolio_metrics = df.groupby('DateTime_$date', as_index=False, group_keys=False).apply(
    lambda x: pd.Series({
        'Portfolio_Value': (x['DEFAULT_LTP'].astype(float) * x['DEFAULT_LotSize'].astype(float)).sum()
    })
).reset_index()

# Handling missing or zero values in 'Portfolio_Value' by replacing them with NaN
portfolio_metrics['Portfolio_Value'] = portfolio_metrics['Portfolio_Value'].replace(0, float('nan'))

# Applying linear interpolation to fill missing values for smooth visualization
portfolio_metrics['Portfolio_Value'] = portfolio_metrics['Portfolio_Value'].interpolate(method='linear')

# Creating an interactive plot to visualize portfolio value over time
fig = go.Figure()

# Adding a line plot for portfolio value with smoothing for better visualization
fig.add_trace(go.Scatter(
    x=portfolio_metrics['DateTime_$date'],
    y=portfolio_metrics['Portfolio_Value'],
    mode='lines+markers',  # Showing both lines and markers for clarity
    name='Portfolio Value',
    line=dict(color='#636efa', shape='spline', smoothing=1.3),  # Using spline for smoothness
    marker=dict(size=4, color='teal')  # Setting marker properties
))

# Customizing the layout for better readability and aesthetics
fig.update_layout(
    title='Portfolio Value Over Time',  # Setting the title of the chart
    xaxis_title='Date',  # Label for the x-axis
    yaxis_title='Portfolio Value (₹)',  # Label for the y-axis
    showlegend=True,  # Displaying the legend
    plot_bgcolor='#e5ecf6',  # Setting a light background color
    height=550,  # Defining the height of the chart
    width=1100,  # Defining the width of the chart
    xaxis=dict(
        showgrid=True,
        gridcolor='white',
        showline=False,
        linecolor='rgb(204, 204, 204)'
    ),
    yaxis=dict(
        showgrid=True,
        gridcolor='white',
        showline=False,
        linecolor='rgb(204, 204, 204)'
    ),
    margin=dict(l=15, r=15, t=80, b=50),  # Increase top margin for legend space
    legend=dict(
        orientation='h',  # Horizontal legend
        x=0.5, y=1.12,  # Center the legend below the title
        xanchor='center',
    )
)

# Displaying the final portfolio value if data is available
if not portfolio_metrics.empty:
    final_value = portfolio_metrics.iloc[-1]['Portfolio_Value']
    final_date = portfolio_metrics.iloc[-1]['DateTime_$date'].strftime('%Y-%m-%d')
    print(f"Final Portfolio Value as of {final_date}: ₹{final_value:,.2f}")
else:
    print("No portfolio data available.")

# Displaying the plot
fig.show()


Final Portfolio Value as of 2023-08-31: ₹31,628,496.00


C:\Users\rohit\AppData\Local\Temp\ipykernel_1624\539541415.py:17: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



In [17]:
#%pip install statsmodels
#%pip install scikit-learn

<h2 style="color: rgb(31, 136, 217);">Final Portfolio Performance Statistical Analysis</h2>

In [21]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import statsmodels.api as sm
from scipy import stats
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from statsmodels.tsa.holtwinters import ExponentialSmoothing

# Load Portfolio Data
df = filtered_df.copy()
df['DateTime_$date'] = pd.to_datetime(df['DateTime_$date'], errors='coerce')

# Calculate Portfolio Value
df['Portfolio_Value'] = df['DEFAULT_LTP'].astype(float) * df['DEFAULT_LotSize'].astype(float)

# Aggregate Portfolio Value by Date
portfolio_metrics = df.groupby('DateTime_$date', as_index=False)['Portfolio_Value'].sum()

# Convert Dates to Numeric Values
portfolio_metrics['Days'] = (portfolio_metrics['DateTime_$date'] - portfolio_metrics['DateTime_$date'].min()).dt.days
X = portfolio_metrics[['Days']].values
y = portfolio_metrics['Portfolio_Value'].values

# ===== Hypothesis Testing (One-Sample t-test) ===== #
benchmark_value = portfolio_metrics['Portfolio_Value'].mean()
t_stat, p_value = stats.ttest_1samp(portfolio_metrics['Portfolio_Value'], benchmark_value)

# Print Hypothesis Testing Results
print("\nHypothesis Testing Results:")
print(f"t-statistic: {t_stat:.4f}, p-value: {p_value:.4f}")
if p_value < 0.05:
    print("Conclusion: Significant change in portfolio value over time (Reject H₀).")
else:
    print("Conclusion: No significant change in portfolio value over time (Fail to reject H₀).")

# ===== OLS Regression (Linear Trend) ===== #
X_ols = sm.add_constant(portfolio_metrics['Days'])
model_ols = sm.OLS(y, X_ols).fit()
portfolio_metrics['Predicted_OLS'] = model_ols.predict(X_ols)

# Print OLS Regression Summary
print("\nOLS Regression Results:")
print(model_ols.summary())

# ===== Polynomial Regression (Degree = 5) ===== #
poly = PolynomialFeatures(degree=5)
X_poly = poly.fit_transform(X)
poly_model = LinearRegression()
poly_model.fit(X_poly, y)
y_pred_poly = poly_model.predict(X_poly)

# ===== Exponential Smoothing (Time Series Forecasting) ===== #
exp_model = ExponentialSmoothing(y, trend="add", seasonal=None, damped_trend=False, use_boxcox=True).fit(remove_bias=True)
y_pred_exp = exp_model.fittedvalues

# Predict Portfolio Value for Aug 31, 2023
target_date = pd.Timestamp("2023-08-31")
days_since_start = (target_date - portfolio_metrics['DateTime_$date'].min()).days
X_target = poly.transform([[days_since_start]])
predicted_poly_value = poly_model.predict(X_target)[0]
predicted_exp_value = exp_model.forecast(1)[0]

# Calculate R-squared Scores for Model Comparison
r2_poly = r2_score(y, y_pred_poly)
r2_exp = r2_score(y, y_pred_exp)

# ===== Plotly Visualization ===== #
fig = go.Figure()

# **Actual Portfolio Value**
fig.add_trace(go.Scatter(
    x=portfolio_metrics['DateTime_$date'],
    y=portfolio_metrics['Portfolio_Value'],
    mode='lines+markers',
    name='Actual Portfolio Value',
    line=dict(color='#636efa')
))

# **OLS Regression Line**
fig.add_trace(go.Scatter(
    x=portfolio_metrics['DateTime_$date'],
    y=portfolio_metrics['Predicted_OLS'],
    mode='lines',
    name='OLS Regression',
    line=dict(color='teal', dash='dash')
))

# **Polynomial Regression Trend**
fig.add_trace(go.Scatter(
    x=portfolio_metrics['DateTime_$date'],
    y=y_pred_poly,
    mode='lines',
    name='Polynomial Regression (5th Degree)',
    line=dict(color='red', dash='dot')
))

# **Exponential Smoothing Trend**
fig.add_trace(go.Scatter(
    x=portfolio_metrics['DateTime_$date'],
    y=y_pred_exp,
    mode='lines',
    name='Exponential Smoothing',
    line=dict(color='orange', dash='dash')
))

# **Prediction for Aug 31, 2023**
fig.add_trace(go.Scatter(
    x=[target_date, target_date],
    y=[predicted_poly_value, predicted_exp_value],
    mode='markers+text',
    name='Aug 31 Predictions',
    marker=dict(size=10, color=['black', 'purple']),
    text=[f"Poly: ₹{predicted_poly_value:,.2f}", f"Exp: ₹{predicted_exp_value:,.2f}"],
    textposition="bottom right"
))

# Update Layout
fig.update_layout(
    title='Portfolio Value Over Time (Actual vs Predicted)',
    xaxis_title='Date',
    yaxis_title='Portfolio Value (₹)',
    showlegend=True,
    plot_bgcolor='#e5ecf6',
    height=500,
    width=1100,  # Keep overall chart width
    margin=dict(l=15, r=15, t=80, b=50),  # Increase top margin for legend space
    xaxis=dict(domain=[0, 1]),  # Expands the plot area within the figure
    legend=dict(
        orientation='h',  # Horizontal legend
        x=0.5, y=1.08,  # Center the legend below the title
        xanchor='center',
    )
)

# Display Predictions
print(f"Polynomial Regression Predicted Value for Aug 31, 2023: ₹{predicted_poly_value:,.2f} (R²: {r2_poly:.4f})")
print(f"Exponential Smoothing Predicted Value for Aug 31, 2023: ₹{predicted_exp_value:,.2f} (R²: {r2_exp:.4f})")

# Show Plot
fig.show()



Hypothesis Testing Results:
t-statistic: 0.0000, p-value: 1.0000
Conclusion: No significant change in portfolio value over time (Fail to reject H₀).

OLS Regression Results:
                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.112
Model:                            OLS   Adj. R-squared:                  0.104
Method:                 Least Squares   F-statistic:                     14.86
Date:                Wed, 12 Mar 2025   Prob (F-statistic):           0.000189
Time:                        23:04:13   Log-Likelihood:                -2148.3
No. Observations:                 120   AIC:                             4301.
Df Residuals:                     118   BIC:                             4306.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err   

C:\Users\rohit\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\statsmodels\tsa\holtwinters\model.py:918: ConvergenceWarning:

Optimization failed to converge. Check mle_retvals.



### Portfolio Value Statistical Analysis - Summary of Findings  

### 1. Hypothesis Testing (One-Sample t-test)  
- The t-statistic is 0.0000, and the p-value is 1.0000, which is much higher than 0.05.  
- This indicates no statistically significant change in portfolio value over time.  
- Conclusion: We fail to reject the null hypothesis (H₀), meaning the portfolio value has remained relatively stable.  

### 2. OLS Regression (Linear Trend Analysis)  
- The R² value is 0.112, meaning only 11.2% of the variation in portfolio value is explained by time.  
- The coefficient for Days (90,040) suggests a positive trend, but the low R² indicates a weak predictive ability.  
- Conclusion: There is a statistically significant upward trend (p < 0.05), but the model has limited explanatory power.  

### 3. Polynomial Regression (5th Degree)  
- The R² value is 0.5611, meaning 56.11% of the variation in portfolio value is captured by the model.  
- The polynomial regression predicted the portfolio value for August 31, 2023, as ₹17,443,014.70.  
- Conclusion: This model provides a better fit compared to OLS but may still be prone to overfitting.  

### 4. Exponential Smoothing (Time Series Forecasting)  
- The R² value is 0.5976, indicating that 59.76% of the variation is explained by this model.  
- The predicted portfolio value for August 31, 2023, is ₹32,441,572.37.  
- Conclusion: Exponential smoothing provides the best fit among the models tested, making it the most reliable for short-term forecasting.  

### 5. Overall Insights  
- The hypothesis test suggests no significant change in portfolio value over time.  
- The linear trend shows a weak upward movement but lacks strong explanatory power.  
- Polynomial regression and exponential smoothing models perform better, with exponential smoothing showing the highest predictive accuracy. 
